runtime/activations/activations.go

```diff
@@ -21,44 +21,63 @@
 //
 package activations
 
-import (
-	"github.com/raviqqe/hamt"
-
-	"github.com/onflow/cadence/runtime/common"
-)
-
-// An Activation is an immutable map of strings to arbitrary values.
+// An Activation is a map of strings to arbitrary values.
 // It can be used to represent an active scope in a program,
 // i.e. it can be used as a symbol table during semantic analysis,
 // or as an activation record during interpretation.
 //
-type Activation hamt.Map
+type Activation struct {
+	entries map[string]interface{}
+	Depth   int
+	Parent  *Activation
+}
 
-func NewActivation() Activation {
-	return Activation(hamt.NewMap())
+func NewActivation(parent *Activation) *Activation {
+	var depth int
+	if parent != nil {
+		depth = parent.Depth + 1
+	}
+	return &Activation{
+		entries: map[string]interface{}{},
+		Depth:   depth,
+		Parent:  parent,
+	}
 }
 
 // Find returns the value for a given key in the activation.
 // It returns nil if no value is found.
 //
-func (a Activation) Find(name string) interface{} {
-	return hamt.Map(a).Find(common.StringEntry(name))
+func (a *Activation) Find(name string) interface{} {
+	result, ok := a.entries[name]
+	if ok {
+		return result
+	}
+
+	if a.Parent != nil {
+		return a.Parent.Find(name)
+	}
+
+	return nil
 }
 
-// Insert inserts the given key-value pair into the activation.
+// Set sets the given key-value pair in the activation.
 //
-func (a Activation) Insert(name string, value interface{}) Activation {
-	return Activation(hamt.Map(a).Insert(common.StringEntry(name), value))
+func (a *Activation) Set(name string, value interface{}) {
+	a.entries[name] = value
 }
 
 // ForEach calls the given function for each entry (key-value pair) in the activation.
 // It can be used to iterate over all entries of the activation.
 //
-func (a Activation) ForEach(f func(string, interface{}) error) error {
-	return hamt.Map(a).ForEach(func(entry hamt.Entry, v interface{}) error {
-		name := string(entry.(common.StringEntry))
-		return f(name, v)
-	})
+func (a *Activation) ForEach(cb func(string, interface{}) error) error {
```

Comment:
This doesn't look like it iterates over the entries in the parent. This will be tricky, because I assume you don't want to iterate over any entires in a parent that are "shadowed" in a descendent. Not sure how to do this efficiently, to be honest!

Reply:
Refactored in [`56e851c` (#571)](https://github.com/onflow/cadence/pull/571/commits/56e851c1ecfd496bc815e5fa6e21d3362936b5a8)